THREESCALE-11536: remove `rspec_api_documentation`

jlledom opened the pull request:
**Description**:

This removes the `rspec_api_documentation` gem. It's used to generate some documentation for the backend API from a few tests suites written using a particular DSL.

I'm not sure where is this documentation used/published/needed... I never heard of it before. The gem is a problem because it's abandoned and can't be updated. Check the description in https://github.com/3scale/apisonator/pull/411

**Issue**:

https://issues.redhat.com/browse/THREESCALE-11536

**Notes for reviewers**:

This PR consists on removing the gem and migration the affected test suites to the regular rspec syntax, basically:

- move `header` calls to inside a `before` callback
- replace `resource` by `describe`
- replace `example` and `example_request` by `it`
- Launch the requests directly through method names (`get`, `post`, etc), instead of calling `do_request` (under `example`) or making the call implicit (under `example_request`)
- remove calls to `parameter`, they were intended for documentation purposes and to generate the request from inside `example_request`. Not needed in  regular rspec.

akostadinov commented:
I even don't want to review everything in details. It looks good from what I saw. One question though is, can this be easily conerted to some other lib that generates openapi docs? Or at least tag the tests somehow, that they are kind of minimal tests that can be used for apidocs generation in case we want to use such a library in the future?

Apisonator might be a good testbed in case we want to do something similar with porta.

jlledom commented:
> One question though is, can this be easily conerted to some other lib that generates openapi docs? Or at least tag the tests somehow, that they are kind of minimal tests that can be used for apidocs generation in case we want to use such a library in the future?

I don't know of any library able to understand `rspec_api_documentation` DSL. I was looking for one a couple of months ago when I was working on the sinatra update and couldn't find one.

`rspec_api_documentation` can be used to generate openapi jsons: https://github.com/zipmark/rspec_api_documentation?tab=readme-ov-file#open_api

So we could generate them now and add them to the PR if they are good enough. If we make changes to the backend API later on, we'll have to update the openapi jsons manually

akostadinov commented:
IIRC looking at something like this in the past, there were a couple of recent libraries that used rspec dsl to generate api docs. So my question was whether migrating to one of these, instead of generic rspec would be easy enough. Or somehow mark the tests so it can be done at a laterpoint.

jlledom commented:
> Question: is the gem really abandoned? There is a whole bunch of commits from 2024.

We have been discussing this after the sync, it got some updates the first quarter last year, but nothing since 10 months ago, also no releases since 2018, not supporting OAS3, etc... and forcing us to maintain a fork because it doesn't support rack 3.

mayorova commented on spec/spec_helper.rb:
```diff
@@ -2,8 +2,8 @@
-# This fixes a `NoMethodError` error caused by `rspec_api_documentation`, which is fixed since 2022, but never released.
-# Ref: https://github.com/zipmark/rspec_api_documentation/commit/758c879893a21233c0eb977e79ef026f263fc37e
+require 'active_support'
+require 'active_support/core_ext/object/json'
```
so, requiring `active_support` isn't pulling the `core_ext` dependencies?

jlledom replied:
Apparently it doesn't. Tests fail if I remove any of the three `active_support` requires.

https://guides.rubyonrails.org/active_support_core_extensions.html

jlledom commented:
@akostadinov I removed the rake task: https://github.com/3scale/apisonator/pull/419/commits/7ef07fec2b6a26cd4f6692b96e2a2845a149f763

@mayorova I pushed the static OAS file: https://github.com/3scale/apisonator/pull/419/commits/e6122dc7dd742b67fea3f43992f93d93a3896e65

I made some small changes like renaming some endpoints to remove the `(Prefix: ...)` in the title, for consistency, because some endpoints had it and some others didn't

mayorova commented on spec/acceptance/api/internal/alert_limits_spec.rb:
```diff
@@ -1,94 +1,100 @@
+      let(:service_id)  { '7575' }
 
-    let(:service_id)  { '7575' }
-    let(:raw_post) { params.to_json }
```
For reference:
https://github.com/zipmark/rspec_api_documentation/blob/d32bf17932a379052e8bb7540fc29ce7e72712ae/README.md#raw_post

jlledom replied:
Thanks for the info 👍

mayorova commented:
This looks great, I would suggest though to move the `backend_internal.yaml` to another folder.
I find `active_docs` misleading, because this spec will not actually be used for the admin portal's active docs. Moreover, in another PR I would remove the `Service Management API.json`, because it's obsolete. Maybe we can put this yaml to `docs/specs` or something like this.

jlledom commented:
@mayorova a few points:

- I removed the fake endpoint
- endpoints `status` and `check.json` are fine. I though they wer out of the `internal` namespace but it turns out they don't. Because there are both `/status` and `/internal/status`, same for `check.json`
- everything under `/internal/` needs to  authorized by HTTP Basic Auth. I added the changes to the yaml but after a few hours I'm starting to think SwaggerUI has a bug, because it considers the yaml valid, it finds the auth and it adds the header on the auto-generated `curl` it shows as example, but it doesn't send the header under the hood.

> This looks great, I would suggest though to move the `backend_internal.yaml` to another folder. I find `active_docs` misleading, because this spec will not actually be used for the admin portal's active docs. Moreover, in another PR I would remove the `Service Management API.json`, because it's obsolete. Maybe we can put this yaml to `docs/specs` or something like this.

OK

mayorova commented:
> * I'm starting to think SwaggerUI has a bug, because it considers the yaml valid, it finds the auth and it adds the header on the auto-generated `curl` it shows as example, but it doesn't send the header under the hood.

I think it's not SwaggerUI, but the JetBrains plugin, because in https://editor.swagger.io/ the Authorization header is sent as expected (but it has an issue with CORS when trying to hit a locally running apisonator :grimacing:)

mayorova reviewed (approved):
Looks great to me :+1:

akostadinov commented:
It makes no sense to address localhost from a remote site... there might be certain protections. How about trying to hit whatever public accessible http server?

jlledom commented:
> It makes no sense to address localhost from a remote site... there might be certain protections. How about trying to hit whatever public accessible http server?

When using the RubyMine plugin, we are querying `localhost` from `localhost`. So no CORS problem here. I guess that's why I don't see the `OPTIONS` query in this scenario. However, the credentials are still not sent

akostadinov commented:
How is the request being performed? some js? or as a normal form submission?

jlledom commented:
> How is the request being performed? some js? or as a normal form submission?

I tried three ways to send the request:

1. Broswer: https://editor.swagger.io/
![image](https://github.com/user-attachments/assets/f23b2312-53c6-4efa-badb-e568553046bd)
2. From RubyMine:
![image](https://github.com/user-attachments/assets/4d7c6f42-3084-46db-a328-a77baa54b1ed)
3. From console:
```
$ curl -X 'GET'   'http://localhost:3001/internal/status'   /
       -H 'accept: */*'   /
       -H 'Authorization: Basic c3lzdGVtX2FwcDpwYXNzd29yZA=='

{"status":"ok","version":{"backend":"3.4.3"}}
```

mayorova commented:
@jlledom @akostadinov Yes, it can easily be confirmed that the requests sent by the plugin to not contain the proper Authorization header by updading the spec URL to some echo API (despite the generated curl that does have the proper Authorization header)

As for CORS, inded, I think different browsers may show the CORS flow differently, I think it doesn't matter.
I enabled CORS on the internal API on my locally running apisonator, and it all works fine from editor.swagger.io.

So, no issue (just that the plugin is slightly less usable than I was hoping for)

mayorova commented on docs/specs/backend_internal.yaml:
```diff
@@ -0,0 +1,939 @@
+  title: 3scale Backend Internal API
+  version: 1.0.0
+servers:
+- url: http://localhost:3000/internal
```
```suggestion
- url: http://localhost:3001/internal
```

Because normally backend runs on `3001`, while `3000` is used for porta

mayorova replied:
But we can leave it until further spec improvements.

jlledom replied:
I pushed the change

mayorova commented:
> So, no issue (just that the plugin is slightly less usable than I was hoping for)

There is actually a workaround. If apisonator is started with empty `CONFIG_INTERNAL_API_USER` and `CONFIG_INTERNAL_API_PASSWORD` values, then the internal API will not have authentication.

jlledom commented:
> As for CORS, inded, I think different browsers may show the CORS flow differently, I think it doesn't matter. I enabled CORS on the internal API on my locally running apisonator, and it all works fine from editor.swagger.io.
> 
> So, no issue (just that the plugin is slightly less usable than I was hoping for)

Thanks for checking!
